src/NHibernate/SqlCommand/SqlSelectBuilder.cs

```diff
@@ -300,17 +303,23 @@ public SqlString ToSqlString()
 
 		private string GetForUpdateString()
 		{
-			if (!Dialect.SupportsOuterJoinForUpdate && hasOuterJoin)
-			{
-				if (Equals(lockMode, LockMode.Upgrade))
-					return Dialect.GetForUpdateString(aliasToLock);
-				if (Equals(lockMode, LockMode.UpgradeNoWait))
-					return Dialect.GetForUpdateNowaitString(aliasToLock);
-			}
+			if (Dialect.SupportsOuterJoinForUpdate || !HasOuterJoin())
+				return Dialect.GetForUpdateString(lockMode);
+
+			if (Equals(lockMode, LockMode.Upgrade))
+				return Dialect.GetForUpdateString(mainTableAlias);
+			if (Equals(lockMode, LockMode.UpgradeNoWait))
+				return Dialect.GetForUpdateNowaitString(mainTableAlias);
 
 			return Dialect.GetForUpdateString(lockMode);
-		}
 
+			bool HasOuterJoin()
+			{
+				return
+					StringHelper.ContainsCaseInsensitive(fromClause, "outer join") ||
+					!string.IsNullOrWhiteSpace(outerJoinsAfterFrom?.ToString());
```

Comment:
It's better to swap those conditions and if whitespace check is really required use `SqlString.IsEmptyOrWhitespace`
https://github.com/nhibernate/nhibernate-core/blob/f8b1bc46f92c7a8cd503d3450f73f6b33ad4e8fb/src/NHibernate/SqlCommand/SqlString.cs#L658

Reply:
@Doan-Van-Tuan could you please incorporate this suggestion?

Reply:
I'll resume work on this fix in next few days. It was our Lunar New Year.